### PR TITLE
Update signpost styleguide (readme)

### PR DESCRIPTION
**Pull request by adressel**
This PR's goal is to discuss and then update the README in Signpost's javascript style guide. While there are other files from the merge with airbnb/master, let's please focus on discussing the file README.md on the top level of this repository.
In a subsequent PR, I intend to update the linters with the rules we agree on.

Thanks to Matt E for giving me some of his inputs!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,921 +79,2324 @@
...
+  const g = function g() {};
+
+  // better
+  // this implicitly sets the function's name to 'h'
+  const h = () => {}
+  ```
 
-      anonymous(); // => TypeError anonymous is not a function
+  <a name="functions--mutate-params"></a><a name="7.12"></a>
+  - [7.12](#functions--mutate-params) Never mutate parameters. eslint: [`no-param-reassign`](http://eslint.org/docs/rules/no-param-reassign.html)
```
Comment (adressel) [inline README.md]: We're certainly not consistent with this rule, but we should be...

File: README.md [hunk 2]
```diff
@@ -72,921 +79,2324 @@
...
+
+    console.log(a); // undefined
+    console.log(b); // undefined
+    console.log(c); // undefined
+
+    // the same applies for `const`
+    ```
+
+  <a name="variables--unary-increment-decrement"></a><a name="13.6"></a>
+  - [13.6](#variables--unary-increment-decrement) Avoid using unary increments and decrements (++, --). eslint [`no-plusplus`](http://eslint.org/docs/rules/no-plusplus)
```
Comment (adressel) [inline README.md]: I don't have a strong preference regarding this one, and would be fine with getting rid of this rule.

File: README.md [hunk 3]
```diff
@@ -72,921 +79,2324 @@
...
 
-  - For more information see [Truth Equality and JavaScript](http://javascriptweblog.wordpress.com/2011/02/07/truth-equality-and-javascript/#more-2108) by Angus Croll
-
-**[⬆ back to top](#table-of-contents)**
-
-
-## Blocks
-
-  - Use braces with all multi-line blocks.
+  <a name="whitespace--around-keywords"></a><a name="18.
```
Comment (adressel) [inline README.md]: I use spaces, but I'd be ok with omitting them in the future.

Comment (adressel) [inline README.md]: I left this in here to have an open discussion. bleeding eyes vs. nicer git diffs

File: README.md [hunk 4]
```diff
@@ -1396,162 +2760,267 @@
...
     ```
 
+  <a name="testing--for-real"></a><a name="28.2"></a>
+  - [28.2](#testing--for-real) **No, but seriously**:
+   - Whichever testing framework you use, you should be writing tests!
+   - Strive to write many small pure functions, and minimize where mutations occur.
+   - Be cautious about stubs and mocks - they can make your tests more brittle.
+   - We primarily use [`mocha`](https://www.npmjs.com/package/mocha) at Airbnb. [`tape`](https://www.npmjs.com/package/tape) is also used occasionally for small, separate modules.
+   - 100% test coverage is a good goal to strive for, even if it's not always practical to reach it.
+   - Whenever you fix a bug, _write a regression test_. A bug fixed without a regression test is almost certainly going to break again in the future.
```
Comment (adressel) [inline README.md]: This point is hardly doable since we move quite fast - leaving it up for discussion.

File: .idea/javascript.iml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (sarahleeashraf) [inline .idea/javascript.iml]: should .idea be in the gitignore?

File: README.md [hunk 6]
```diff
@@ -72,921 +79,2324 @@
...
     // good
-    var items = [];
+    const obj = {
+        lukeSkywalker,
+    };
     ```
 
-  - If you don't know array length use Array#push.
+  <a name="objects--grouped-shorthand"></a><a name="3.7"></a>
+  - [3.5](#objects--grouped-shorthand) Group your shorthand properties at the beginning of your object declaration.
```
Comment (rpaterson) [inline README.md]: Not a fan of this - all other things being equal I would prefer to have the props in alphabetical order regardless of which use the shorthand.

File: README.md [hunk 7]
```diff
@@ -72,921 +79,2324 @@
...
-    var goSportsTeam = true;
-    var dragonball = 'z';
+    const foo = function bar() {
+        // ...
+    };
     ```
 
-  - Declare unassigned variables last. This is helpful when later on you might need to assign a variable depending on one of the previous assigned variables.
+  <a name="functions--in-blocks"></a><a name="7.2"></a>
+  - [7.2](#functions--in-blocks) Never declare a function in a non-function block (if, while, etc). Assign the function to a variable instead. Browsers will allow you to do it, but they all interpret it differently, which is bad news bears. eslint: [`no-loop-func`](http://eslint.org/docs/rules/no-loop-func.html)
```
Comment (rpaterson) [inline README.md]: Most of our code does not run in-browser - if that's really the only reason for this, do we need it in our general style guidelines?

File: README.md [hunk 8]
```diff
@@ -72,921 +79,2324 @@
...
-      //..other stuff..
+    // good
+    function foo(name, options, args) {
+        // ...
+    }
+    ```
 
-      var name = getName();
+  <a name="es6-rest"></a><a name="7.5"></a>
+  - [7.5](#es6-rest) Never use `arguments`, opt to use rest syntax `...` instead. eslint: [`prefer-rest-params`](http://eslint.org/docs/rules/prefer-rest-params)
```
Comment (rpaterson) [inline README.md]: We should use this in the `static create()` methods we add to enable Monkey-Patcher.
```Javascript
static create(...args) { return new MyModule(...args); }
```

File: README.md (hunk 1 above)
Comment (rpaterson) [inline README.md]: There used to be an argument for doing this to "return" multiple values, but now that JS has object destructuring that excuse is gone.

File: README.md [hunk 9]
```diff
@@ -72,921 +79,2324 @@
...
+
+    // good
+    function f2(obj) {
+        const key = Object.prototype.hasOwnProperty.call(obj, 'key') ? obj.key : 1;
     }
     ```
 
-  - Named function expressions hoist the variable name, not the function name or the function body.
+  <a name="functions--reassign-params"></a><a name="7.13"></a>
+  - [7.13](#functions--reassign-params) Never reassign parameters. eslint: [`no-param-reassign`](http://eslint.org/docs/rules/no-param-reassign.html)
```
Comment (rpaterson) [inline README.md]: This one's gonna hurt.  We do this everywhere to handle default values.

Comment (sp) [inline README.md]: That's why the JS gods gave us default parameter syntax, right?

File: README.md [hunk 10]
```diff
@@ -72,921 +79,2324 @@
...
+
+    // good
+    [1, 2, 3].map((x) => {
+        const y = x + 1;
+        return x * y;
+    });
+    ```
+
+  <a name="arrows--implicit-return"></a><a name="8.2"></a>
+  - [8.2](#arrows--implicit-return) If the function body consists of a single expression, omit the braces and use the implicit return. Otherwise, keep the braces and use a `return` statement. eslint: [`arrow-parens`](http://eslint.org/docs/rules/arrow-parens.html), [`arrow-body-style`](http://eslint.org/docs/rules/arrow-body-style.html) jscs:  [`disallowParenthesesAroundArrowParam`](http://jscs.info/rule/disallowParenthesesAroundArrowParam), [`requireShorthandArrowFunctions`](http://jscs.info/rule/requireShorthandArrowFunctions)
```
Comment (akallem) [inline README.md]: Should we still do this even if the single expression is very long, e.g., a call to `Sequencer`. Personal preference, but I think the implicit return is most useful when the expression itself is short. If the expression is long, explicitly creating a block and a `return` doesn't add much weight (relative to the expression) and I think makes it clearer.

File: README.md [hunk 11]
```diff
@@ -72,921 +79,2324 @@
...
+  <a name="modules--use-them"></a><a name="10.1"></a>
+  - [10.1](#modules--use-require) Import modules with `require`.
+
+    ```javascript
+    // good
+    const StyleGuide = require('./SignpostStyleGuide');
+    ```
+
+  <a name="modules--indentation"></a><a name="10.2"></a>
+  - [10.2](#modules--indentation) Import all modules at the beginning of a file and line up the `require`s.
```
Comment (akallem) [inline README.md]: The lining up of requires can sometimes lead to larger than necessary diffs (e.g., if a new module is required that's bigger than the existing indentation). I think we should discuss whether this slightly nicer looking format is really worth it.

File: README.md [hunk 12]
```diff
@@ -72,921 +79,2324 @@
...
 
     // bad
-    var isJedi = luke['jedi'];
+    const errorMessage = 'This is a super long error that was thrown because ' +
+        'of Batman. When you stop to think about how Batman had anything to do ' +
+        'with this, you would get nowhere fast.';
 
     // good
-    var isJedi = luke.jedi;
+    const errorMessage = 'This is a super long error that was thrown because of Batman. When you stop to think about how Batman had anything to do with this, you would get nowhere fast.';
```
Comment (JLeySignpost) [inline README.md]: So we prefer longer strings over linter rules that enforce length? I get that ideally we just don't have lines over 100, but that isn't always as easily said as done

File: README.md (hunk 11 above)
Comment (adressel) [inline README.md]: Right, which is why I'd suggest adding some extra spaces for safety. I usually start the `=` 35 characters into the line.

File: README.md (hunk 10 above)
Comment (adressel) [inline README.md]: I'll change this rule slightly. I think that arrow functions should always be used without curly braces if possible. We should discuss whether or not to include parentheses around a single argument in this case (as I know that there are different opinions on the team).

File: README.md (hunk 12 above)
Comment (adressel) [inline README.md]: There are mainly three arguments that come to my mind in this discussion:
1) Pro: It's nice if long strings are searchable.
2) Con: It might be hard to read the strings in an IDE.
3) Con: You'd have to disable the linter rule whenever your string is more than 100/120 characters long (which is possible for single lines in your code, but slightly ugly).

File: README.md (hunk 7 above)
Comment (adressel) [inline README.md]: I don't have a strong preference regarding this rule. Could you point me to an example where we declare a function in a non-function block or where this would be useful?

File: README.md (hunk 9 above)
Comment (rpaterson) [inline README.md]: Absolutely, I'm just saying we're going to have to change this in a lot of places.

File: README.md (hunk 7 above)
Comment (rpaterson) [inline README.md]: Not off the top of my head.  This comes up occasionally when using old skool for-loops instead of `Array.forEach()` (so we can exit early)

File: README.md (hunk 10 above)
Comment (rpaterson) [inline README.md]: I definitely think we should _not_ include parens around a single argument.  I don't see the benefit, and it's more terse.

Comment (rpaterson) [inline README.md]: As it says below, "Less visual clutter"

File: README.md [hunk 13]
```diff
@@ -72,921 +79,2324 @@
...
+    // good
+    class PeekableQueue extends Queue {
+        peek() {
+            return this.queue[0];
+        }
+    }
+    ```
+
+  <a name="constructors--chaining"></a><a name="9.2"></a>
+  - [9.2](#constructors--chaining) Methods can return `this` to help with method chaining.
```
Comment (rpaterson) [inline README.md]: This is in effect making a DSL.  It's tempting and I'm guilty of doing it myself (e.g. toshokan/QueryBuilder) but I think it's probably not a good idea most of the time.

File: README.md (hunk 11 above)
Comment (rpaterson) [inline README.md]: I'm usually a big fan of creative indentation for vertical alignment, but there doesn't seem to be much benefit of doing it in the requires section because the name of the `const` variable is almost always the same as the name of the Module file - so if you're searching for something it's easy to scan the `const` names which are already vertically aligned, even though the requires are not.

File: README.md [hunk 14]
```diff
@@ -72,921 +79,2324 @@
...
+        dragonball = 'z';
+
+    // good
+    const items = getItems();
+    const goSportsTeam = true;
+    const dragonball = 'z';
+    ```
+
+  <a name="variables--const-let-group"></a><a name="13.3"></a>
+  - [13.3](#variables--const-let-group) Group all your `const`s and then group all your `let`s.
```
Comment (rpaterson) [inline README.md]: Not sure I see how useful this one is - seems like it's more useful to write the code in logical order.

File: README.md [hunk 15]
```diff
@@ -72,921 +79,2324 @@
...
+        if (name === 'test') {
+            this.setName('');
+            return false;
+        }
+
+        return name;
+    }
+    ```
+  <a name="variables--no-chain-assignment"></a><a name="13.5"></a>
+  - [13.5](#variables--no-chain-assignment) Don't chain variable assignments.
```
Comment (rpaterson) [inline README.md]: This one seems wrong to me.  The problem in the example isn't the _assignment_, it's the _declaration_ of the variables.  IMO this should be OK:
```Javascript
let a = null;
let b = null;
// ... later ...
a = b = 7;
```
I often use this for the module class declaration:
```Javascript
const MyModule = module.exports = class MyModule {
    // ...
};
MyModule.SOME_STATIC_CONSTANT = 'abc';
```

File: README.md (hunk 2 above)
Comment (rpaterson) [inline README.md]: I vote get rid of it.  Would be especially annoying in a `for` loop. (Still need to use a `for` loop instead of `forEach()` if you want to exit early.)

File: README.md [hunk 16]
```diff
@@ -72,921 +79,2324 @@
...
+        // ...
+    }
+
+    // good
+    if (isValid) {
+        // ...
+    }
+
+    // bad
+    if (name) {
```
Comment (rpaterson) [inline README.md]: Is this so bad?

File: README.md [hunk 17]
```diff
@@ -72,921 +79,2324 @@
...
+        // ...
+    }
```
Comment (rpaterson) [inline README.md]: Is this so bad?

File: README.md [hunk 18]
```diff
@@ -72,921 +79,2324 @@
...
+
+## Blocks
+
+  <a name="blocks--braces"></a><a name="16.1"></a>
+  - [16.1](#blocks--braces) Use braces with all multi-line blocks.
+
+    ```javascript
+    // bad
+    if (test)
+        return false;
```
Comment (rpaterson) [inline README.md]: I would be OK with allowing this explicitly in the case where the code inside the `if` block is only one line.  Less visual clutter.  I find it annoying to add brackets around a single line.

File: README.md [hunk 19]
```diff
@@ -72,921 +79,2324 @@
...
+    // bad
+    if (test) return false;
+
+    // good
+    if (test) {
+        return false;
+    }
+
+    // bad
+    function foo() { return false; }
```
Comment (rpaterson) [inline README.md]: Similar to my comment above, I think we should allow this if it's only one statement in the block - less visual clutter.

File: README.md [hunk 20]
```diff
@@ -72,921 +79,2324 @@
...
+    ```
+
+
+**[⬆ back to top](#table-of-contents)**
+
+
+## Comments
+
+  <a name="comments--multiline"></a><a name="17.1"></a>
+  - [17.1](#comments--multiline) Use `/** ... */` for multi-line comments.
```
Comment (rpaterson) [inline README.md]: It's probably my Java background, but IMO `/** ... */` is for JSDoc comments, `/* ... */` is for regular multi-line comments in code...

File: README.md [hunk 21]
```diff
@@ -72,921 +79,2324 @@
...
+    // also good
+    function getType() {
+        // set the default type to 'no type'
+        const type = this.type || 'no type';
+
+        return type;
+    }
+    ```
+
+  - [17.3](#comments--spaces) Start all comments with a space to make it easier to read. eslint: [`spaced-comment`](http://eslint.org/docs/rules/spaced-comment)
```
Comment (rpaterson) [inline README.md]: I started using this recently in the warehouse project.  It was annoying at first and took a while to train my fingers, but I like it now.

File: README.md (hunk 3 above)
Comment (rpaterson) [inline README.md]: The argument for this one is the same reason we should put a comma after the last element in a multi-line array or object declaration...

Comment (rpaterson) [inline README.md]: I find this one really annoying, and we just removed it from Glaze.  Can we get rid of this?  A blank line after a function declaration is nice - gives the code room to breathe.

Comment (rpaterson) [inline README.md]: Yes!  Better diffs!

File: README.md [hunk 22]
```diff
@@ -995,384 +2405,338 @@
...
     ```
 
 **[⬆ back to top](#table-of-contents)**
 
 
 ## Naming Conventions
 
-  - Avoid single letter names. Be descriptive with your naming.
+  <a name="naming--descriptive"></a><a name="22.1"></a>
+  - [22.1](#naming--descriptive) Avoid single letter names. Be descriptive with your naming. eslint: [`id-length`](http://eslint.org/docs/rules/id-length)
```
Comment (rpaterson) [inline README.md]: Can we add an exception for `i`, `j`, and `k` in for loops?  (ESLint has a way to make exceptions to this rule just for those names).

File: README.md (hunk 4 above)
Comment (rpaterson) [inline README.md]: I think this is a fine goal.  Automated tests are faster than manual QA.

File: README.md (hunk 7 above)
Comment (adressel) [inline README.md]: Could you give a full example where you can't assign the function to a variable?

File: README.md (hunk 11 above)
Comment (adressel) [inline README.md]: I see two arguments for it:
1) It looks better
2) It IS easier to find modules since all the requires are far to the right and don't disturb the eye when reading.

However, I'd like to extend this rule and enforce importing the modules in alphabetical order.

File: README.md (hunk 13 above)
Comment (adressel) [inline README.md]: Right, I don't think we use this very often - the rule says `can` though, and I think it's fine to allow it in cases where it's useful (i.e. when writing a test express response object to test a call like `res.status(200).json({})`)

File: README.md (hunk 14 above)
Comment (adressel) [inline README.md]: By logical order, do you mean declare variables in the order in which they're used? An argument for grouping `const`s and `let`s is that it looks consistent and is easier to read.

File: README.md (hunk 15 above)
Comment (adressel) [inline README.md]: That's correct, I will change the rule to advise against variable `declarations`.

File: README.md (hunk 17 above)
Comment (adressel) [inline README.md]: I'm a bit torn about this one. Generally, I think that this rule can prevent bugs - for example in the case where an empty string would satisfy the condition you _intended_. It would force the developer to actively think about every corner case. A downside is, that you'd have to explicitly go through each case ( `name !== null && name != undefined && name != ''`, in some cases possibly a test that the variable contains a string `typeof name === 'string'`).

File: README.md (hunk 18 above)
Comment (adressel) [inline README.md]: Not how we do it right now, but I'd be open to it. We should also discuss the actual one-line version:
```javascript
if (condition) return 'something';
```

File: README.md (hunk 19 above)
Comment (adressel) [inline README.md]: Open to it, but I think that in almost all cases, you could replace this with an arrow function (which, I know, is not what this rule addresses).

File: README.md (hunk 20 above)
Comment (adressel) [inline README.md]: Right, some people might also prefer using `//` for comments that span over a few lines.

File: README.md (hunk 22 above)
Comment (adressel) [inline README.md]: How often are these loops used in other repos? I was under the impression that they should be avoided if possible (I know you can't break out of them). I'm open to allowing `i`, `j` and `k` for loops, but that would allow using these shorthands anywhere. An alternative would be to deactivate the linter rule when you run into a situation where you need a for loop.

File: README.md (hunk 3 above)
Comment (adressel) [inline README.md]: This is something I'd like to discuss with everyone.

File: README.md (hunk 7 above)
Comment (rpaterson) [inline README.md]: It's not that you _can't_ - you always can.  It's just sometimes annoying or unnatural to do so, and if there's no benefit in our environment then I don't see the benefit of this rule.

File: README.md (hunk 11 above)
Comment (rpaterson) [inline README.md]: +1 for alphabetical order

File: README.md (hunk 14 above)
Comment (rpaterson) [inline README.md]: Yes the order that they're used, or the order that makes sense in the context of the algorithm.  The order that you introduce the characters in a story you tell - you don't necessarily group them by the ones that are going to die at the end and the ones that are going to survive ;-)

File: README.md (hunk 18 above)
Comment (rpaterson) [inline README.md]: Yes I like that too.

File: README.md (hunk 19 above)
Comment (rpaterson) [inline README.md]: Oh sure - +1 for arrow functions

File: README.md (hunk 22 above)
Comment (rpaterson) [inline README.md]: ¯\_(ツ)_/¯ it comes up from time to time

File: README.md (hunk 13 above)
Comment (rpaterson) [inline README.md]: That's an interesting example - in that case you're not writing greenfield code - you're writing test code that mocks code written by someone else.  Should we generally have different rules for test code vs prod code to account for this sort of thing?  ESLint has a hierarchical config file system that makes this easy to implement - e.g. [see here in the warehouse project](https://github.com/signpost/warehouse/blob/master/test/.eslintrc.js#L30).

File: README.md (hunk 11 above)
Comment (adressel) [inline README.md]: Andrew brought up a good point: we use variables to store paths that are commonly used when importing modules (for example `const CoreApiDir = './lib/api';` in core). It would make sense to base them off of a variable `rootDir` which always points to the root directory of a repository. Besides consistency, another pro is that this will make it easier to copy dependencies from file to file.

For example:
```javascript
const rootDir    = '../../..'; // omitting the last '/' on purpose
const appsDir  = `${rootDir}/apps/`;
```

File: README.md (hunk 2 above)
Comment (melyanow) [inline README.md]: you can use for..of to exit early

File: README.md (hunk 6 above)
Comment (melyanow) [inline README.md]: I'm not sure I agree, especially when you have nested objects and the like, shorthand properties can be easy to miss if they're in between two more complicated properties.

File: README.md (hunk 18 above)
Comment (melyanow) [inline README.md]: if we're going to not put brackets around single line `if` blocks, I'd definitely advocate for putting them on the same line, otherwise it's easy to accidentally add a second line and not realize it's outside of the block

File: README.md (hunk 11 above)
Comment (melyanow) [inline README.md]: for alphabatizing, do you guys prefer
a. alphabatize everything
```
const rootDir = '../../..';
const CONFIG  = require('config');
const Log = `${rootDir}/lib/Log`;
const P = require('toshokan/P');
```

b. split into node_modules, libs/helpers from the current repo, and alphabatize each separately
```
const CONFIG  = require('config');
const P = require('toshokan/P');

const rootDir = '../../..';
const Log = `${rootDir}/lib/Log`;
```

Also, for CoreApiClient requires, if I need more than one, currently I generally do 
```
const CoreApiClient = require('toshokan/CoreApiClient');
   const CoreAgencies  = CoreApiClient.Agencies;
   const CoreLocations = CoreApiClient.Locations;
```
is there a preferred way to handle that case?  (I guess one option we could do const `{CoreAgencies, CoreLocations} = require('toshokan/CoreApiClient')`), or do a separate require for each (and leave off the `CoreApiClient= ...`)

File: README.md [hunk 23]
```diff
@@ -72,921 +79,2324 @@
...
+  const module3     = require('./Module3');
+  // ... code unrelated to imports
+  ```
+
+**[⬆ back to top](#table-of-contents)**
+
+## Iterators and Generators
+
+  <a name="iterators--nope"></a><a name="11.1"></a>
+  - [11.1](#iterators--nope) Don't use iterators. Prefer JavaScript's higher-order functions instead of loops like `for-in` or `for-of`. eslint: [`no-iterator`](http://eslint.org/docs/rules/no-iterator.html) [`no-restricted-syntax`](http://eslint.org/docs/rules/no-restricted-syntax)
```
Comment (melyanow) [inline README.md]: Completely disagree with this.  I don't 100% understand the argument about immutability, and `for..of` loops are easier to read (need less nesting), can be broken out of, and is more general, being usable on any iteratable object.

File: README.md (hunk 2 above)
Comment (rpaterson) [inline README.md]: Good point, although I think there's another rule in here somewhere that we're not supposed to use `for..of`

Comment (melyanow) [inline README.md]: yeah, I commented on that too, don't agree with that rule at all

File: README.md (hunk 23 above)
Comment (melyanow) [inline README.md]: Also, the built-in iterators are super useful.  Using Sets is much cleaner than objects or Array.find for anything where you need `O.has(...)`, and being able to use non-strings/Symbols as keys in maps is also very useful.  (Maps in general have a number of ways they're useful over objects, even with strings as keys.  They remove the need to loop over `Object.keys(...)`, and you can spread them easily to get an entries, keys or values array, if you need to do any array.prototype functions on them, which is certainly cleaner than doing `Object.keys(O).map(  key => { value = O[key] }).reduce(...)`, for example.

**Comment by melyanow** [comment]
A few minor things not mentioned at all in the style guide: 
Should we have any specific style-guide rules surrounding the use of Symbols, Proxies/Reflect, Array.includes (as a replacement for .indexOf)? 

Also, not sure if it's in the scope of a style guide, but should we have a section talking about any default modules that should be used across the board for consistency (e.g. using ('toshokan/P') for promises)

File: README.md (hunk 11 above)
Comment (adressel) [inline README.md]: I don't have a strong preference here, and I'm open to discuss it. I usually put the common paths at the very top, followed by an empty line, followed by all imports in alphabetical order (depending on the variable's name, not the import path). I also use one import per dependency, but I'm not strongly opposed to using destructuring in this case.

Another point on this: some people use lower-case variable names for third party dependencies. We should fine a common ground here for the sake of consistency.

File: README.md (hunk 18 above)
Comment (adressel) [inline README.md]: Yes, that makes sense. How about this: if it can fit into one line while still being readable, it's fine to omit the brackets and put the whole statement in one line. Whenever multiple lines are needed, use brackets.

**Comment by adressel** [comment]
We barely use Symbols, Proxies/Reflect. I'd add rules in a separate PR if everyone is fine with that.

I removed one rule advising against Object.assign - what did you have in mind? I'm all for using Array.includes.

**Comment by melyanow** [comment]
yeah, that makes sense.  I didn't have anything specific for any of those, more were just new features I noticed weren't mentioned, so figured it was worth discussion.  
One possible thing for `Object.assign` would be using it only to create new objects, not to mutate existing ones (Object.assign({},...)

File: README.md (hunk 11 above)
Comment (rpaterson) [inline README.md]: I use `lowerCamelCase` variable names for dependencies that export a top-level function, like `dedent`.  Otherwise if it's a module that has static methods and/or gets instantiated then I prefer `UpperCamelCase`.

File: README.md (hunk 3 above)
Comment (adressel) [inline README.md]: I assume this points at rule 18.8, and I'm fine with removing it 👍

Comment (adressel) [inline README.md]: 19.2?

Comment (rpaterson) [inline README.md]: This is referring to `[18.5](#whitespace--newline-at-end) End files with a single newline character`

Comment (rpaterson) [inline README.md]: Yes `[18.8](#whitespace--padded-blocks) Do not pad your blocks with blank lines.`

Comment (rpaterson) [inline README.md]: Yes `[19.2](#commas--dangling) Additional trailing comma`.  I was agreeing with the README on the next line where it says `Why? This leads to cleaner git diffs.`

**Comment by rpaterson** [comment]
Do we have a Git style guide?  I think we need one if we don't.

**Comment by adressel** [comment]
I don't think we do. It's out of the scope of this guide, but it would be good to have some rules.

File: README.md [hunk 24]
```diff
@@ -22,549 +27,1490 @@
...
+
+    console.log(foo[0], bar[0]); // => 9, 9
+    ```
+
+**[⬆ back to top](#table-of-contents)**
+
+## References
+
+  <a name="references--prefer-const"></a><a name="2.1"></a>
+  - [2.1](#references--prefer-const) Use `const` for all of your references; avoid using `var`. eslint: [`prefer-const`](http://eslint.org/docs/rules/prefer-const.html), [`no-const-assign`](http://eslint.org/docs/rules/no-const-assign.html)
```
Comment (akallem) [inline README.md]: Minor nit: This should probably say "Prefer `const`...". We still want to allow `let` (as mentioned below).

File: README.md [hunk 25]
```diff
@@ -22,549 +27,1490 @@
...
+  // good
+  const g = function g() {};
+
+  // better
+  // this implicitly sets the function's name to 'h'
+  const h = () => {}
+  ```
+
+  <a name="functions--mutate-params"></a><a name="7.11"></a>
+  - [7.11](#functions--mutate-params) Never mutate parameters. eslint: [`no-param-reassign`](http://eslint.org/docs/rules/no-param-reassign.html)
```
Comment (akallem) [inline README.md]: I think "never" is too strong here (though I agree it should be heavily discouraged). For example, decorating `req` (which we already do) would violate this rule. But decorating `req` (or any object, for that matter) seems OK to me as long as its commented clearly.

(Side note: this example doesn't really corresponds to the `no-param-reassign` rule.)

**Comment by adressel** [comment]
I will merge the PR this afternoon unless someone has objections.

File: README.md [hunk 26]
```diff
@@ -22,549 +27,1490 @@
...
+
+    ```javascript
+    // bad
+    function sayHi(name) {
+        return 'How are you, ' + name + '?';
+    }
+
+    // bad
+    function sayHi(name) {
+        return ['How are you, ', name, '?'].join();
```
Comment (tmargheim) [inline README.md]: 😱

File: README.md [hunk 27]
```diff
@@ -22,549 +27,1490 @@
...
+    }
+
+    // bad
+    function sayHi(name) {
+        return `How are you, ${ name }?`;
```
Comment (tmargheim) [inline README.md]: Shouldn't the description above mention "don't put spaces around the interpolated value"?

File: README.md [hunk 28]
```diff
@@ -22,549 +27,1490 @@
...
+
+    ```javascript
+    // bad
+    function concatenateAll() {
+        const args = Array.prototype.slice.call(arguments);
+        return args.join('');
+    }
+
+    // good
+    function concatenateAll(...args) {
```
Comment (tmargheim) [inline README.md]: NTH: An example with more leading params, e.g. `(param1, param2, ...args)`

File: README.md [hunk 29]
```diff
@@ -22,549 +27,1490 @@
...
     // good
-    var errorMessage = 'This is a super long error that was thrown because ' +
-      'of Batman. When you stop to think about how Batman had anything to do ' +
-      'with this, you would get nowhere fast.';
+    const StyleGuide = require('./SignpostStyleGuide');
     ```
 
-  - When programmatically building up a string, use Array#join instead of string concatenation. Mostly for IE: [jsPerf](http://jsperf.com/string-vs-array-concat/2).
+  <a name="modules--order"></a><a name="10.2"></a>
+  - [10.2](#modules--order) Import all modules at the beginning of a file and order them alphabetically based on the variable name. Optionally, line up the `require`s for better readability.
```
Comment (tmargheim) [inline README.md]: We should be explicit about delayed `require`s to avoid circular includes -- either they're an exception to this rule, or we're saying those must be avoided even at the cost of restructuring existing code.

File: README.md [hunk 30]
```diff
@@ -651,342 +1725,654 @@
...
-    var totalScore = '' + this.reviewScore;
+    const totalScore = this.reviewScore + ''; // invokes this.reviewScore.valueOf()
 
     // bad
-    var totalScore = '' + this.reviewScore + ' total score';
+    const totalScore = this.reviewScore.toString(); // isn't guaranteed to return a string
 
     // good
-    var totalScore = this.reviewScore + ' total score';
+    const totalScore = String(this.reviewScore);
```
Comment (tmargheim) [inline README.md]: This isn't what the text recommendation says...

File: README.md [hunk 31]
```diff
@@ -651,342 +1725,654 @@
...
     ```
 
-  - Use `parseInt` for Numbers and always with a radix for type casting.
+  <a name="coercion--numbers"></a><a name="21.3"></a>
+  - [21.3](#coercion--numbers) Numbers: Use `Number` for type casting and `parseInt` always with a radix for parsing strings. eslint: [`radix`](http://eslint.org/docs/rules/radix)
```
Comment (tmargheim) [inline README.md]: Interesting! I've been preferring Number()

File: README.md [hunk 32]
```diff
@@ -995,384 +2381,338 @@
...
+    function foo() {
+        return () => {
+            console.log(this);
+        };
     }
     ```
 
-  - Name your functions. This is helpful for stack traces.
+  <a name="naming--filename-matches-export"></a><a name="22.6"></a>
+  - [22.6](#naming--filename-matches-export) A base filename should exactly match the name of its export.
```
Comment (tmargheim) [inline README.md]: This is notable! We're rejecting the `__` convention? Good.

File: README.md (hunk 29 above)
Comment (dreinharth) [inline README.md]: I would say they should be avoided even at the cost of restructuring existing code

File: README.md (hunk 12 above)
Comment (dreinharth) [inline README.md]: This comes up all the time in tests.  How are we going enforce this?

File: README.md (hunk 30 above)
Comment (dreinharth) [inline README.md]: Why did our string coercion advice change?  It was `'' + string` in the past

File: README.md [hunk 33]
```diff
@@ -995,384 +2381,338 @@
...
-    var age = 0;
+    const age = 0;
 
     // bad
-    var hasAge = new Boolean(age);
+    const hasAge = new Boolean(age);
 
     // good
-    var hasAge = Boolean(age);
+    const hasAge = Boolean(age);
```
Comment (dreinharth) [inline README.md]: I think we already do the "best" thing, so why even open the door to this usage?

File: README.md (hunk 12 above)
Comment (rpaterson) [inline README.md]: ESLint has a cascading directory-based config file system that could allow us to specify different rules for the tests - that's what `warehouse` currently does.

File: README.md (hunk 32 above)
Comment (dreinharth) [inline README.md]: That change was part of the original styleguide -- the `__` files predate that advice

File: README.md (hunk 27 above)
Comment (adressel) [inline README.md]: I'll add a note.

File: README.md (hunk 29 above)
Comment (adressel) [inline README.md]: I'm with David here - circular dependencies are _usually_ a sign of poorly factored code.

File: README.md (hunk 30 above)
Comment (adressel) [inline README.md]: I took the change over from the airbnb style guide because I found their argument compelling:

```javascript
// bad
const totalScore = this.reviewScore + ''; // invokes this.reviewScore.valueOf()
```

File: README.md (hunk 31 above)
Comment (adressel) [inline README.md]: Do you mean Number.parseInt? Number() is already mentioned, and I'll add an example for Number.parseInt

File: README.md (hunk 33 above)
Comment (adressel) [inline README.md]: Ok, I'm fine with only allowing the "best" solution.

File: README.md (hunk 28 above)
Comment (adressel) [inline README.md]: This is an example advising against `arguments`. Leading parameters would defeat its purpose, since `...args` should be the equivalent of using `Array.prototype.slice.call(arguments)` in this case.
